### PR TITLE
Build bucket v2 migration add v2 endpoints (Part 4)

### PR DESCRIPTION
**Pull request by ricardoamador**
Adding updated v2 api endpoints. These endpoints will handle the new build bucket v2 api requests. Eventually we will remove the v1 api endpoints once they are no longer processing messages.

*List which issues are fixed by this PR. You must list at least one issue.*
Part 4 of https://github.com/flutter/flutter/issues/135934

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: app_dart/lib/src/request_handlers/postsubmit_luci_subscription_v2.dart [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  // No need to update task in datastore if
+  // 1) the build is `scheduled`. Task is marked as `In Progress`
+  //    whenever scheduled, either from scheduler/backfiller/rerun. We need to update
+  //    task in datastore only for
+  //    a) `started`: update info like builder number.
+  //    b) `completed`: update info like status.
+  // 2) the task is already completed.
+  //    The task may have been marked as completed from test framework via update-task-status API.
+  bool _shouldUpdateTask(bbv2.Build build, firestore.Task task) {
+    return build.status != bbv2.Status.SCHEDULED && !Task.finishedStatusValues.contains(task.status);
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/postsubmit_luci_subscription_v2.dart:172]:
```suggestion
    return build.status != bbv2.Status.SCHEDULED && !firestore.Task.finishedStatusValues.contains(task.status);
```

File: app_dart/lib/src/request_handlers/presubmit_luci_subscription_v2.dart [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    // Skip status update if we can not get the sha tag.
+    if (tags.where((element) => element.key == 'buildset').isEmpty) {
+      log.warning('Buildset tag not included, skipping Status Updates');
+      return Body.empty;
+    }
+
+    log.fine('Setting status (${build.status.toString()}) for $builderName');
+
+    if (pubSubCallBack.hasUserData()) {
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/presubmit_luci_subscription_v2.dart:85]:
```suggestion
    if (!pubSubCallBack.hasUserData()) {
      return Body.empty;
    }
```

https://dcm.dev/docs/rules/common/avoid-nested-conditional-expressions/

File: app_dart/lib/src/request_handlers/presubmit_luci_subscription_v2.dart [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      Map<String, dynamic> userDataMap = <String, dynamic>{};
+      try {
+        log.info('User data was not base64 encoded.');
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/presubmit_luci_subscription_v2.dart:88]: Move below line 89.

File: app_dart/lib/src/request_handlers/scheduler/scheduler_request_subscription.dart [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    return Body.empty;
+  }
+
+  /// Returns [List<bbv2.BatchRequest_Request>] of requests that need to be retried.
+  Future<List<bbv2.BatchRequest_Request>> _sendBatchRequest(bbv2.BatchRequest request) async {
+    log.info('Sending batch request for ${request.toProto3Json().toString()}');
+
+    // Seems to be that the create is necessary here in order to be able to assign a
+    // response here. Similar to instantiation.
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/scheduler/scheduler_request_subscription.dart:96]: Do you mean it doesn't work with 
`final bbv2.BatchResponse response = await buildBucketClient.batch(request);`? What error do yo get?

File: app_dart/lib/src/request_handlers/scheduler/scheduler_request_subscription.dart [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      log.severe('Exception making batch Requests.');
+      rethrow;
+    }
+
+    log.info('Made ${request.requests.length} and received ${response.responses.length}');
+    log.info('Responses: ${response.responses}');
+
+    // By default, retry everything. Then remove requests with a verified response.
+    // THese are the requests in the batch request object. Just requests.
+    final List<bbv2.BatchRequest_Request> retry = request.requests;
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/scheduler/scheduler_request_subscription.dart:110]: Is the null check (from old API) not necessary? I am not sure. But to be safe, shall we keep the logic same?
```
    final List<Request> retry = request.requests ?? <Request>[];
```

File: app_dart/lib/src/request_handlers/dart_internal_subscription.dart [hunk 6]
```diff
@@ -47,17 +49,14 @@ class DartInternalSubscription extends SubscriptionHandler {
       return Body.empty;
     }
 
-    final dynamic buildData = json.decode(message.data!);
-    log.info('Build data json: $buildData');
-
-    if (buildData['build'] == null) {
-      log.info('no build information in message');
-      return Body.empty;
-    }
```
[the remark this answers is not in the document]
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/dart_internal_subscription.dart:56]: Hmm, we do need this. I have added it back.

File: app_dart/lib/src/request_handlers/dart_internal_subscription.dart [hunk 7]
```diff
@@ -76,28 +75,31 @@ class DartInternalSubscription extends SubscriptionHandler {
...
+    getBuildRequest.id = buildId;
 
     log.info(
-      'Calling buildbucket api to get build data for build $buildbucketId',
+      'Calling buildbucket api to get build data for build $buildId',
     );
-    final Build build = await buildBucketClient.getBuild(request);
+
+    bbv2.Build existingBuild = bbv2.Build.create();
+    existingBuild = await buildBucketV2Client.getBuild(getBuildRequest);
```
[the remark this answers is not in the document]
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/dart_internal_subscription.dart:88]: Committed to local branch.

File: app_dart/lib/src/request_handlers/github/webhook_subscription.dart [hunk 8]
```diff
@@ -68,14 +70,19 @@ class GithubWebhookSubscription extends SubscriptionHandler {
     required super.cache,
     required super.config,
     required this.scheduler,
+    required this.schedulerV2,
     required this.gerritService,
     required this.commitService,
     this.githubChecksService,
+    this.githubChecksServiceV2,
```
[the remark this answers is not in the document]
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/github/webhook_subscription.dart]: Removed. Yeah this is odd.

File: app_dart/lib/src/request_handlers/postsubmit_luci_subscription_v2.dart (hunk 1 above)
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/postsubmit_luci_subscription_v2.dart:172]: Diff is stale but I have added the change. Good catch.

File: app_dart/lib/src/request_handlers/presubmit_luci_subscription_v2.dart (hunk 3 above)
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/presubmit_luci_subscription_v2.dart:88]: Done. Updated the other classes with this decode message as well.

File: app_dart/lib/src/request_handlers/presubmit_luci_subscription_v2.dart (hunk 2 above)
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/presubmit_luci_subscription_v2.dart:85]: Refactored to remove 1 level of cyclomatic complexity.

File: app_dart/lib/src/request_handlers/scheduler/scheduler_request_subscription.dart (hunk 4 above)
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/scheduler/scheduler_request_subscription.dart:96]: Ah this must have been from something I was trying a while ago as this was one of the first classes converted. This should be no issue and I will remove the note.

File: app_dart/lib/src/request_handlers/scheduler/scheduler_request_subscription.dart (hunk 5 above)
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/scheduler/scheduler_request_subscription.dart:110]: Yeah that field in the BatchRequest v2 api is not nullable.

**Review by keyonghan — APPROVED** [review]
LGTM! :crossed_fingers::crossed_fingers::crossed_fingers:

**Pull request merged**